system-helper/flatpak-system-helper.c

```diff
@@ -221,6 +806,40 @@ handle_deploy (FlatpakSystemHelper   *object,
       return TRUE;
     }
 
+  ongoing_pull = get_ongoing_pull_by_dir (arg_repo_path);
```

Comment:
you seem to assume here that get_ongoing_pull_by_dir returns a ref ?

Reply:
Yes. As you would have guessed see-ing the g_autoptr-ized `ongoing_pull` variable.

There was a downstream suggestion to remove the pull from the array(under lock) in `get_ongoing_pull_by_dir`. Maybe we can try that? However, I need to re-work idle exit around it then.

Reply:
The premise for the above approach is coming from how `get_ongoing_pull_by_dir` is used. It's called during `handle_deploy` **or** `handle_cancel_pull`. Basically, this should be **ideally** called at most once (hence remove it from the array right there and avoid any potential race). This approach will more look like ownership-transfer to `handle_deploy` for the pull struct and that `handle_deploy` takes care of cleaning it up.

And on the other hand..
I need to ask if there is some wrong/non-obvious thing here. Can't the `handle_deploy` freeup the `ongoing_pull` returned using automatic cleanup?

Reply:
What I intended to suggest was that `get_ongoing_pull_by_dir` should remove the element it returns from the list. The intent was that only the first call to Deploy() or CancelPull() for a given dir would find the element and operate on it. As it stands, that's not the case if the two method calls are made simultaneously (since they're handled in different threads).

Reply:
Slightly getting confused now.

Yeah, so we should take a ref in `get_ongoing_pull_by_dir` right? As mclasen suggested above.

Reply:
Personally, I think that `get_ongoing_pull_by_dir` should:

* if it finds an entry in the list:
  * remove it from the list (without unreffing it)
  * return that pointer (transferring ownership to the caller)
* otherwise:
  * return NULL

Reply:
Ok, I'll probably go with @wjt advice.
Reason: It very much makes sense for current scenario. Either one(Deploy or Cancel) should operate on it and we should eliminate all possibilities for further operation by another thread on the same pointer.
* Renaming `get_ongoing_pull_by_dir` to `take_ongoing_pull_by_dir`
* I did some thinking/reading on idle exit of systemhelper. The big disaster to avoid is exiting the helper in between of `GetChildRepoForPull` and `Deploy or Cancel` invocations. But as we now remove ongoing_pull from array in one of `Deploy or Cancel` invocation,  I think the current logic should be fine in `idle_timeout_cb`. It will not exit as one of these invocation has already has a ref on FlatpakSystemHelper.

Reply:
> GetChildRepoForPull -> GetChildRepoForPull .. for the case of partial pulls ?

That is just a private method used inside GetChildRepoForPull. 
While others are dbus-invoked ones in which a single `ongoing_pull` is meant to be shared/operated on.

> Can we have documentation on expected call sequences somewhere ?
In the commit message?

Reply:
@matthiasclasen  So I think, I mis-understood your comment. You are asking about expected call sequences right?
> GetChildRepoForPull -> GetChildRepoForPull for partial pulls...
This doesn't look right to me. Calling twice `GetChildRepoForPull` without `CancelPull` or `Deploy` should be a invalid use-case. Reason being, the `user_dir` is still mounted at that point of time. `umount` is required (either by `Deploy` or `CancelPull`) Does it answer your question? :thinking: